src/supervisor.tex

```diff
@@ -2261,3 +2281,108 @@ \chapter{``Svnapot'' Standard Extension for NAPOT Translation Contiguity, Versio
+enabled to observe different memory attributes for a given region of
+physical memory than a concurrent access to the same page performed by M-mode
+or when {\tt satp}.MODE=Bare.  In such cases, each individual access observes
+the memory attributes associated with its own path; aliases are not considered
+when determining attributes.
+
+\begin{commentary}
+For example, a cacheable access may be issued at the same time as a
+non-cacheable access to the same physical memory address.  In this case,
+if the former is performed first in the global memory order, then it will
+be evicted from the cache by the latter.  If on the other hand the cacheable
+access appears after the non-cacheable access, then the former may remain
+cached as it normally would.
+
+Likewise, accesses performed under memory indicating the non-idempotent
+attribute must not be merged with idempotent accesses to the same region
+in flight at the same time, as the non-idempotency of the former must
+be respected.  This is not expected to be a common situation.
+
+Note that Svpbmt cannot be used to completely prevent speculative reads from
+being performed to a region of memory for which the PMAs indicate idempotence,
+as speculation can still be performed via M-mode or via Bare mappings, which do
+not use the PBMTs.
+\end{commentary}
+
+Memory accesses are ordered according to the effective memory type for the
+address in question.  A region of memory which is considered main memory by the
+PMAs but I/O by a PTE will obey channel 0 strong ordering, where for memory
+ordering purposes the ``I/O region'' is considered to be main memory.  In other
+words, each access to such a virtual address will appear in the global memory
+order after all prior operations from the same hart to main memory or to the
+same ``I/O region'', and it will appear earlier in the global memory order than
+all subsequent accesses from the same hart to main memory or to the same ``I/O
+region''.  Such operations are not guaranteed to remain ordered with respect to
+I/O operations to other I/O regions unless FENCEs are inserted by software.
+
+Likewise, a region of memory which is considered I/O by the PMAs but main
+memory by a PTE will obey RVWMO (or RVTSO if enabled) rather than I/O
+strong ordering rules, and accesses to such pages are considered main
+memory rather than I/O for the purposes of FENCE, {\em.aq}, and {\em.rl}.
+In such cases it is the responsibility of software to insert FENCEs
+or other ordering mechanisms if necessary.
+
+\begin{commentary}
+A device driver written to rely on I/O strong ordering rules will not
+operate correctly if the address range is mapped as main memory by the
+page-based memory types.  Operating systems and hypervisors must take
+care to apply such a combination only when strong ordering is not
+actually needed.
+
+In spite of the caveat above, it will often still be useful to map
+device memory regions in I/O as main memory so that write combining
```

Comment:
To some readers this can be confusing or ambiguous since it isn't clear whether "I/O" and "main memory" and "device memory regions" are referring to architectural I/O and main memory regions, or to colloquial physical main memory or memory-mapped I/O.

This is also misleading since it doesn't matter that an address is mapped to colloquial physical DRAM main memory.  the effective memory type will determine how the access is handled/performed.

Reply:
How about "...it will still be useful to map physical I/O regions containing physical device memory into effective main memory regions..."